### PR TITLE
Reduce objmode and unpickling overhead

### PR DESCRIPTION
**Pull request by sklam**
Depends on https://github.com/numba/numba/pull/6017
Closes https://github.com/numba/numba/pull/6024

Proposed patch to reduce objmode function overhead and unpickling overhead.

- https://github.com/numba/numba/commit/831d5bf62dfa20006d68c05a3673cf0b4e6c754e Fixes objmode function unnecessarily unpickling the type object of `OmittedArg`, which is only needed for function that has default arguments.
- https://github.com/numba/numba/commit/2a6ac69f61058f5d0671679ff39bdb83a55ebf69 keeps a memo (a `dict`) to cache unpickled object by C `numba_unpickle()` to reduce overhead at unpickling at runtime. Keeping a memo allows us to clear the cached objects to reduce memory consumption.

**Background of the runtime unpickling**

Numba pickle python objects and reconstructs them at runtime for certain operations (like boxing, unboxing) that needs to reference live python objects. But a compiled function cannot hold references to live objects directly; otherwise, it can't be cached. Pickling let's us get around that problem by serializing live objects into a byte string that we can reconstruct at runtime.

### CODE REVIEW

**Comment by sklam** [comment]
There are some scary bugs & segfault

**Comment by sklam** [comment]
There's a reproducible segfault with ` numba.tests.test_lists.TestLists.test_index3`

**Comment by sklam** [comment]
Segfault is happening at https://github.com/python/cpython/blob/4bfcffe16e9742c154f54ae96b5b36903500abaa/Objects/typeobject.c#L3306

```
Program received signal SIGSEGV, Segmentation fault.
0x00005555556e0bdd in type_dealloc () at /tmp/build/80754af9/python_1565725737370/work/Objects/typeobject.c:3306
3306	/tmp/build/80754af9/python_1565725737370/work/Objects/typeobject.c: No such file or directory.
(gdb) bt
#0  0x00005555556e0bdd in type_dealloc () at /tmp/build/80754af9/python_1565725737370/work/Objects/typeobject.c:3306
#1  0x0000555555644f45 in _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at /tmp/build/80754af9/python_1565725737370/work/Python/ceval.c:1842
#2  0x00005555556b7ccb in function_code_fastcall (globals=<optimized out>, nargs=4, args=<optimized out>, 
    co=<optimized out>) at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:283
#3  _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:408
#4  0x000055555571fa93 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1565725737370/work/Python/ceval.c:4616
#5  _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1565725737370/work/Python/ceval.c:3110
#6  0x00005555556b7ccb in function_code_fastcall (globals=<optimized out>, nargs=1, args=<optimized out>, 
    co=<optimized out>) at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:283
#7  _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:408
#8  0x000055555571f806 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1565725737370/work/Python/ceval.c:4616
#9  _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1565725737370/work/Python/ceval.c:3124
#10 0x0000555555668539 in _PyEval_EvalCodeWithName ()
    at /tmp/build/80754af9/python_1565725737370/work/Python/ceval.c:3930
```

**Comment by sklam** [comment]
the problem is likely to be in `numba_do_raise`

**Comment by sklam** [comment]
`numba_do_raise` is too complicated. i think the logical thing to do is to refactor it to simplify the controlflow and the refcount management. Error in refcount is the cause for the segfault; i.e. dropping Py_DECREF in that function "fixes" the segfault.

**Comment by sklam** [comment]
attempting and testing fix at https://github.com/numba/numba/pull/6017

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/6024 tested this PR combined with https://github.com/numba/numba/pull/6017 which fixes the segfault seen here.

**Comment by stuartarchibald** [comment]
> #6024 tested this PR combined with #6017 which fixes the segfault seen here.

Manually tested, it certainly doesn't segfault when combined.

File: numba/core/callconv.py [hunk 1]
```diff
@@ -316,8 +316,9 @@ def get_exception(self, exc_id):
             msg = "unknown error %d in native function" % exc_id
             return SystemError, (msg,)
 
-
-excinfo_t = ir.LiteralStructType([GENERIC_POINTER, int32_t])
+# The structure type constructed by PythonAPI.serialize_uncached()
+# i.e a {i8* pickle_buf, i32 pickle_bufsz, i8* hash_buf}
```
Comment (stuartarchibald) [inline numba/core/callconv.py:320]: The docstring for `serialize_uncached` needs updating, it still refers to the variant without the hash. https://github.com/numba/numba/pull/5922/files#diff-e3116b002a61bdf2a13aff3dda7ed96fR1313

File: numba/core/pylowering.py [hunk 2]
```diff
@@ -170,6 +167,15 @@ def lower_inst(self, inst):
         else:
             raise NotImplementedError(type(inst), inst)
 
+    @utils.cached_property
+    def _omitted_typobj(self):
+        """Return a `OmittedArg` type instance in LLVM value suitable for
```
Comment (stuartarchibald) [inline numba/core/pylowering.py:172]:
```suggestion
        """Return a `OmittedArg` type instance as a LLVM value suitable for
```

File: numba/core/serialize.py [hunk 3]
```diff
@@ -133,6 +133,32 @@ def _rebuild_code(marshal_version, bytecode_magic, marshalled):
     return marshal.loads(marshalled)
 
 
+# Keep unpickled object via `numba_unpickle` alive.
+_unpickled_memo = {}
+
+
+def _numba_unpickle(address, bytedata, hashed):
```
Comment (stuartarchibald) [inline numba/core/serialize.py:140]: Is there merit in adding a couple of tests just to make sure memoization is working as expected?

File: numba/core/callconv.py (hunk 1 above)
Comment (sklam) [reply numba/core/callconv.py:320]: updated in https://github.com/numba/numba/pull/5922/commits/ebbae83720e6c29ef6fbb756298211d24869279f

File: numba/core/pylowering.py (hunk 2 above)
Comment (sklam) [reply numba/core/pylowering.py]: updated in https://github.com/numba/numba/pull/5922/commits/ebbae83720e6c29ef6fbb756298211d24869279f

File: numba/core/serialize.py (hunk 3 above)
Comment (sklam) [reply numba/core/serialize.py:140]: added a simple test in https://github.com/numba/numba/pull/5922/commits/17e72af86d927aa5d01bd7f1fbb80f8b036b0797 for the memoization of its output

File: numba/core/pylowering.py [hunk 4]
```diff
@@ -170,6 +167,15 @@ def lower_inst(self, inst):
         else:
             raise NotImplementedError(type(inst), inst)
 
+    @utils.cached_property
+    def _omitted_typobj(self):
+        """Return a `OmittedArg` type instance as a LLVM value suitable for
+        testing at runtime.
+        """
+        from numba.core.dispatcher import OmittedArg
```
Comment (stuartarchibald) [inline numba/core/pylowering.py:175]: this here due to circular import problems?

File: numba/core/pylowering.py [hunk 5]
```diff
@@ -188,21 +194,29 @@ def lower_assign(self, inst):
         elif isinstance(value, ir.Yield):
             return self.lower_yield(value)
         elif isinstance(value, ir.Arg):
+            # TODO: check why value.name may not exist
```
Comment (stuartarchibald) [inline numba/core/pylowering.py:197]: This TODO apply still? What's the context that triggers it?

File: numba/core/pylowering.py (hunk 4 above)
Comment (sklam) [reply numba/core/pylowering.py:175]: yes, confirmed. can't move it to module scope

File: numba/core/pylowering.py (hunk 5 above)
Comment (sklam) [reply numba/core/pylowering.py:197]: It no longer applies. To be removed.

Comment (sklam) [reply numba/core/pylowering.py]: removed in https://github.com/numba/numba/pull/5922/commits/d328ac507d7c05fa731e856972f22344e7151b34

**Pull request merged**